### PR TITLE
feat: add Node Selector and Tolerations

### PR DESCRIPTION
**Pull request by vimalk78**
Fixes: https://github.com/sustainable-computing-io/kepler-operator/issues/130
Fixes POWERMON-118

### CODE REVIEW

File: pkg/api/v1alpha1/kepler_types.go [hunk 1]
```diff
@@ -17,6 +17,7 @@ limitations under the License.
 package v1alpha1
 
 import (
+	v1 "k8s.io/api/core/v1"
```
Comment (sthaha) [inline pkg/api/v1alpha1/kepler_types.go:20]:
```suggestion
	corev1 "k8s.io/api/core/v1"
```
To follow the convention

File: pkg/components/exporter/exporter.go [hunk 2]
```diff
@@ -99,17 +107,10 @@ func NewDaemonSet(k *v1alpha1.Kepler) *appsv1.DaemonSet {
...
 					DNSPolicy:          corev1.DNSPolicy(corev1.DNSClusterFirstWithHostNet),
-					Tolerations: []corev1.Toleration{
-						{
-							Effect: corev1.TaintEffectNoSchedule, Key: "node-role.kubernetes.io/master",
-						},
-						{
-							Effect: corev1.TaintEffectNoSchedule, Key: "node-role.kubernetes.io/infra",
-						},
-					},
+					Tolerations:        append(defaultTolerations, tolerations...),
```
Comment (sthaha) [inline pkg/components/exporter/exporter.go:113]: I am not really sure of  ... I think we should apply the  toleration **only** if the user has not specified a toleration.
What if I want my deployment to run on normal nodes , i.e. do not run on master (for whatever reason).

File: pkg/components/exporter/exporter_test.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+func TolerationsFromDS(ds *appsv1.DaemonSet) []corev1.Toleration {
+	return ds.Spec.Template.Spec.Tolerations
+}
+
+func TestNodeSelection(t *testing.T) {
+
+	tt := []struct {
+		spec     ko.ExporterSpec
```
Comment (sthaha) [inline pkg/components/exporter/exporter_test.go:23]: could you please add a scenario? 
```suggestion
		scenario   string
		spec     ko.ExporterSpec
```
And use 
```
t.Run(tc.scenario, func(...){
   t.Parallel()

})
```

File: pkg/components/exporter/exporter_test.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
+package exporter
+
+import (
+	"testing"
+
+	ko "github.com/sustainable.computing.io/kepler-operator/pkg/api/v1alpha1"
```
Comment (sthaha) [inline pkg/components/exporter/exporter_test.go:6]:
```suggestion
	"github.com/sustainable.computing.io/kepler-operator/pkg/api/v1alpha1"
```
To match the convention 👼

File: pkg/components/exporter/exporter_test.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	"gotest.tools/v3/assert"
```
Comment (sthaha) [inline pkg/components/exporter/exporter_test.go:7]: nit: we should rely on just one assert package. I am okay with any :) but pick one and use that.

Comment (vimalk78) [reply pkg/components/exporter/exporter_test.go:7]: is there something already chosen? i will use that.

File: pkg/components/exporter/exporter.go (hunk 2 above)
Comment (vimalk78) [reply pkg/components/exporter/exporter.go:113]: users can add another taint to exclude master nodes in that case

File: pkg/api/v1alpha1/kepler_types.go (hunk 1 above)
Comment (vimalk78) [reply pkg/api/v1alpha1/kepler_types.go:20]: ok

**Review by sthaha — CHANGES_REQUESTED** [review]
Looks good overall, few minor comments to address.

File: pkg/components/exporter/exporter_test.go (hunk 4 above)
Comment (vimalk78) [reply pkg/components/exporter/exporter_test.go:6]: ok

File: pkg/api/v1alpha1/kepler_types.go [hunk 6]
```diff
@@ -98,6 +99,14 @@ type ExporterSpec struct {
...
 	// +kubebuilder:validation:Minimum=1
 	Port int32 `json:"port,omitempty"`
+
+	// Defines which Nodes the Pod is scheduled on
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, define Pod's tolerations
+	// +optional
+	Tolerations []corev1.Toleration `json:"tolerations,omitempty"`
```
Comment (sthaha) [inline pkg/api/v1alpha1/kepler_types.go:109]: Just wondering if we can set the defaults here?

Comment (vimalk78) [reply pkg/api/v1alpha1/kepler_types.go:109]: in code it is fine too, not too much difference.

**Review by sthaha — APPROVED** [review]
Looks great! 
There is a suggestion about setting the defaults for `nodeSelector` and `tolerations` but we can follow this up in a different PR.

**Pull request merged**